kong/resty/dns/client.lua

```diff
@@ -839,7 +813,37 @@ local function syncQuery(qname, r_opts, try_list)
   local key = qname..":"..r_opts.qtype
   local item = queue[key]
 
-  -- if nothing is in progress, we start a new async query
+  local is_yieldable = in_yieldable_phase()
+
+  -- If nothing is in progress and in the yieldable phase,
+  -- we start a new sync query
+
+  if not item and is_yieldable then
+    item = {
+      key = key,
+      semaphore = semaphore(),
+      qname = qname,
+      r_opts = cycle_aware_deep_copy(r_opts),
+      try_list = try_list,
+    }
+    queue[key] = item
+
+    item.result, item.err = individualQuery(qname, item.r_opts, try_list)
+
+    -- query done, but by now many others might be waiting for our result.
+    -- 1) stop new ones from adding to our lock/semaphore
+    queue[key] = nil
+    -- 2) release all waiting threads
+    item.semaphore:post(math_max(item.semaphore:count() * -1, 1))
+    item.semaphore = nil
```

Comment:
I noticed this is copied from `executeQuery`, but it seems we aren't check item.semaphore null-ness before using it on Line 870. Should we add a check?

Reply:
let me check, taking a look into it

Reply:
For the completion, we need to check its null-ness. But the semaphore.new() is very based resource allocation function, i check its implmentation, there is only one error situation where the memory allocation fails. So I think checking its nullness causes code redudancy, which is like checkting lua table new allocation. And in kong source base, most calls of semaphore.new() does not check its return value. And in its official documentation, it does not check its return value, see [here](https://github.com/openresty/lua-resty-core/blob/master/lib/ngx/semaphore.md#synchronizing-threads-in-the-same-context)

Reply:
hi @fffonion I have reviewed the logic, thinking it is unnecessary to check for nullness. Except for the scenario where it can between obtaining the `item` and using the `item.semaphore`, I have checked all instances of this type of context, and I couldnt find any explicit `yield()` or `ngx.sleep`/network-IO operation.